### PR TITLE
[SDK 544] XCFrameworks/SPM branch

### PR DESCRIPTION
**Pull request by alessioborraccino**
## Description

Repackages the framework into a xcframework 

## Motivation and Context

For future proof (as apple is slowly deprecated fat frameworks) 

## How is it being implemented?

* Dependency management changed:
1) Changed dependency management for building the SDK from carthage to swift package 
2) Due to the fact that swift packages are automatically packaged as nested in the main framework (which is disallowed for the appstore), as workaround there is a new static library target packaging all swift pm dependencies into one, and used as a not embedded dependency for the sdk main target.

* Code changed
creating an xcframework requires module stability and enabling library evolution mode. This comported some changes:
1) all the dependencies used internally needed to be marked as @_implementationOnly so that wouldnt appear in the public module interface
2) Some of the public api (identifiers and common Fhir Resource) were not admissible for module stability so they had to be moved into the FHIR Library instead. This means that with this change:
  a) FHIR Library agnostically permits to add any identifier with out API with some helpers on resources, while the SDK with the same public api of before will internally use it to add our Partner ID (namely nothing changes
  b) FHIR Library unifies both Fhir3 and 4 domain resource behind the Marker protocol of anyFHirResource, so that this can be used in the SDK (was before in the SDK and not admissable)
  c) Attachment r4 helper also had to be moved in the fhir library (which to be fair makes more sense)
3) Some extensions of Foundation classes are also not admissable so had to be removed from the public api 
4) Some Equatable/Hashable implementation were removed / moved to test targets because only needed there

* Tests changed
Tests for both frameworks needed to be updated with more direct imports in order to be testable with Swiftpm. With this PR all the changes on tests were enough to run the tests, but emerged a need to change some of the tests in order to be able to run some of them, but will be focus of another PR.

* Other changes
1) Await is no longer a usable word so had to put ticks around it
2) unarchiver migration was no longer needed (as it was referring to a very old version of the sdk 

## How Has This Been Tested?

Imported in a dedicated project on my machine via spm AND via pods, and run.
Tested in the example app for main functionalities.
Ran current unit tests.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have updated the changelog accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: Package.swift [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        .library(
+            name: "Data4LifeSDK",
+            targets: ["Data4LifeSDK", "Data4LifeDependencies"]),
+        .library(
+            name: "Data4LifeCrypto",
+            targets: ["Data4LifeCrypto"]),
+    ],
+    dependencies: [
+        .package(name: "Data4LifeSDKUtils",
+                 url: "git@github.com:d4l-data4life/d4l-utils-ios.git",
```
Comment (ztepsa) [inline Package.swift:21]: Can SDK Utils and FHiR lib also be https links instead of ssh?

File: SDK/Sources/Client/Data4LifeClient.swift [hunk 2]
```diff
@@ -14,8 +14,8 @@
 //  contact D4L by email to help@data4life.care.
 
 import Foundation
```
Comment (ztepsa) [inline SDK/Sources/Client/Data4LifeClient.swift:16]: We can remove `Foundation` as there's an import for `UIKit`.

File: Gemfile.lock [hunk 3]
```diff
@@ -238,17 +254,16 @@ GEM
...
-  asciidoctor
-  cocoapods
-  fastlane
-  synx
+  asciidoctor (= 2.0.12)
+  cocoapods (= 1.10.1)
+  fastlane (= 2.178.0)
 
 BUNDLED WITH
    2.1.4
```
Comment (ztepsa) [inline Gemfile.lock:269]: Maybe upgrade bundler to latest version (`2.2.15`)?

File: SDK/Sources/Models/Record/FHIRRecord/DecryptedFHIRRecord.swift [hunk 4]
```diff
@@ -97,7 +100,7 @@ struct DecryptedFhirR4Record<T: FhirR4Resource>: DecryptedRecord {
             let dataKey = try decryptDataKey(from: encryptedRecord, commonKey: commonKey, cryptoService: cryptoService)
             let attachmentKey = try decryptAttachmentKey(from: encryptedRecord, commonKey: commonKey, cryptoService: cryptoService)
 
-            let resourceProxy = try await(cryptoService.decrypt(data: encryptedData,
+            let resourceProxy = try `await`(cryptoService.decrypt(data: encryptedData,
```
Comment (ztepsa) [inline SDK/Sources/Models/Record/FHIRRecord/DecryptedFHIRRecord.swift:103]: Maybe add typealias for `await` instead of using backtick?

**Review by ztepsa — COMMENTED** [review]
Looks good 🚀

File: Package.swift (hunk 1 above)
Comment (alessioborraccino) [reply Package.swift:21]: I don't know if it works, i had problems making it work before with https (as in xcode would still try to authenticate with git and somehow fail) and i decided to make it a problem for another PR if thats ok.

File: Gemfile.lock (hunk 3 above)
Comment (alessioborraccino) [reply Gemfile.lock:269]: any specific reason?

**Review by jose-cuadros-data4life — CHANGES_REQUESTED** [review]

File: Gemfile.lock (hunk 3 above)
Comment (alessioborraccino) [reply Gemfile.lock]: done

File: SDK/Sources/Models/Record/FHIRRecord/DecryptedFHIRRecord.swift (hunk 4 above)
Comment (alessioborraccino) [reply SDK/Sources/Models/Record/FHIRRecord/DecryptedFHIRRecord.swift]: changed to `wait`

File: SDK/Sources/Services/FHIRService/Protocols/FhirAttachmentOperations.swift [hunk 5]
```diff
@@ -14,9 +14,9 @@
 //  contact D4L by email to help@data4life.care.
 
 import Foundation
-import Data4LifeFHIR
-import Then
-import struct Data4LifeCrypto.Key
+@_implementationOnly import Data4LifeCrypto
+
```
[the remark this answers is not in the document]
Comment (alessioborraccino) [reply SDK/Sources/Services/FHIRService/Protocols/FhirAttachmentOperations.swift]: done

File: SDK/Sources/Client/Data4LifeClient.swift (hunk 2 above)
Comment (alessioborraccino) [reply SDK/Sources/Client/Data4LifeClient.swift]: done

File: Package.swift (hunk 1 above)
Comment (ztepsa) [reply Package.swift:21]: sounds good 👍

**Review by ztepsa — APPROVED** [review]

File: HCSDK.xcodeproj/project.pbxproj [hunk 6]
```diff
@@ -82,6 +82,7 @@
 		3679AAA725923BF90006494F /* Data4LifeClient+FhirR4Tests.swift in Sources */ = {isa = PBXBuildFile; fileRef = 3679AAA625923BF90006494F /* Data4LifeClient+FhirR4Tests.swift */; };
 		3679AAB125923D560006494F /* Data4LifeClient+FhirR4AttachmentTests.swift in Sources */ = {isa = PBXBuildFile; fileRef = 3679AAB025923D560006494F /* Data4LifeClient+FhirR4AttachmentTests.swift */; };
 		3679AAB725923DCB0006494F /* Data4LifeClient+FhirR4DocumentReferenceTests.swift in Sources */ = {isa = PBXBuildFile; fileRef = 3679AAB625923DCB0006494F /* Data4LifeClient+FhirR4DocumentReferenceTests.swift */; };
+		3682F78F2640381E00C03F55 /* Await+Swift6.swift in Sources */ = {isa = PBXBuildFile; fileRef = 3682F78E2640381E00C03F55 /* Await+Swift6.swift */; };
```
Comment (jose-cuadros-data4life) [inline HCSDK.xcodeproj/project.pbxproj:85]: 🎉

**Review by jose-cuadros-data4life — APPROVED** [review]
Nice! I tested example app, tests, fastlane build for xcframeworks, etc.. everything looks good to me 👍

**Pull request merged**